### DISCUSSION TITLE
Putting `electrs` behind auth

### DISCUSSION THREAD

**Discussion by torkelrogstad** [romanz/electrs#926] [Q&A] [answered]
Is it possible to put `electrs` behind auth, i.e. require that clients provide authorization before requests are served?

**Comment by romanz** [comment]
IIUC there is no such feature in the Electrum protocol (as of today).
We may try to add a new command - or change [server.version](https://electrumx-spesmilo.readthedocs.io/en/latest/protocol-methods.html#server.version) behavior to allow only specific clients to connect.
@Kixunil WDYT?

**Reply by Kixunil** [reply]
I don't see any sane way to do it. Tunnels are fine IMO, we just need better support from integrators (yes, that includes myself).

**Comment by antonilol** [comment]
could this be achieved with an electrum server behind a proxy that requires username/password?

**Reply by torkelrogstad** [reply]
That's been my thinking as well, but I'm not sure how to accomplish that with a protocol that relies on raw TCP connections. If this was an HTTP server I could pass something into the headers, but is that even possible with TCP?

**Reply by Kixunil** [reply, marked as answer]
The only way is to use some kind of tunneling - VPN, SSH, Tor, or whatever else you can think of. We have some of them [documented](https://github.com/romanz/electrs/blob/master/doc/config.md). I personally use SSH tunnel because it's simple and supports key authentication - no passwords, every client gets its own key. If you're worried about race conditions and such use a private network namespace in Linux (IDK what about other OSes).

**Reply by dlukt** [reply]
Does electrs need to be publicly accessible?

**Reply by romanz** [reply]
No - I personally suggest running it on a private network.

**Reply by dlukt** [reply]
But Roman, doesn't it need to speak to other Electrum nodes? Or is it just a client to bitcoind middleware? I'm asking because I'm new to the whole thing.
I'm having trouble to connect from my arch desktop box and from android via Wireguard provided VPN address.
Why do you advise for a private network?
Are you afraid there might be security issues you can't forsee and someone might exploit e.g. off by one or similar, or is the consideration more on the DoS side of things, electrs not being able to handle tcp flooding?

**Reply by Kixunil** [reply]
It's just an additional index for your bitcoind. "middleware" might not be entirely accurate but it's pretty close, yeah. There's no DoS protection, so someone could make it stuck or something.

I don't know exactly how Wireguard works but I think you need to set electrs to be bound to that address. E.g. if the address of VPN interface is 10.1.2.3 then you need to set the listen address to 10.1.2.3 (rather than 127.0.0.1 which is the default).

**Reply by dlukt** [reply]
Kixunil, my setup is electrs listens on 127.0.0.1:50001, nginx uses the server's x509 cert from letsenrypt and listens on wireguard_local_address:50002.
That works with Sparrow but not the Electrum client v4.5.8.
So I thought I'll just make the nginx TLS stream wrapper listen on public_IP:50002, since it's using the server's name as CN anyway.
What might be possible is a mismatch of the CN(cannonical name) of the x509 cert and the local IP address, thus Electrum desktop and Android client not connecting and Sparrow being less picky about it.
I'm not too worried about DoS attacks. If someone decides they just have to do a DoS attack I can put it behind the wireguard VPN again, but for simplicity's sake I'd just expose nginx on the public IPs.

Edit: Yep, that was it. After listening on the public IP, Electrum client on the Linux Desktop connects, but not on the Android phone.
I hate BTC and Electrum
And another edit. After disconnecting from WiFi and going 5G the connecting from the phone is successful.

**Reply by Kixunil** [reply]
I suggest you remove the nginx thing, wireguard already encrypts the traffic and the additional component may is just adding complexity and possible problems for no benefit.
